### DISCUSSION TITLE
Image is uploading with errors through google cloud cpp client, but can upload successfully with gsutil

### DISCUSSION THREAD

**Discussion by elamacchia** [googleapis/google-cloud-cpp#7519] [Q&A] [answered]
I haven't found much about specifically uploading images - I'm trying to include this functionality within my c++ code, but not sure if I'm missing something. My upload code looks like this:
```
auto client = gcs::Client();
StatusOr<gcs::ObjectMetadata> metadata = client.UploadFile(
            file_name, bucket_name, target_path,  gcs::ContentType("image/jpeg"));.
```
where file_name is my local file, bucket_name is the bucket name and target_path is the name to save the file as. The upload seems to work correctly (metadata is correct) and generates a file in the bucket with the target name, but clicking on or downloading the actual image says it contains errors. In contrast, when I run `gsutil cp file_name gs://bucket_name`, it uploads correctly without errors.

**Comment by coryan** [comment]
> but clicking on or downloading the actual image says it contains errors. 

What are you using the "click on" or download the file?  And what is the error message?  And could you share the OS you used?  I just wrote a [repro](https://github.com/googleapis/google-cloud-cpp/files/7421006/repro-7519.tar.gz) for this and:

* I was able to see an uploaded (JPEG) image using the GCS console.
* I was also able to download it using:

```
curl -v -H "Authorization: Bearer $(gcloud auth print-access-token)" https://www.googleapis.com/storage/v1/b/${BUCKET_NAME}/o/${OBJECT_NAME}?alt=media -o /dev/null
```

And the `content-type` and other headers look correct to me.

**Reply by elamacchia** [reply]
This is on ubuntu 18.04. I uploaded a screenshot below of the error.

**Comment by elamacchia** [comment]
![Screenshot from 2021-10-26 12-37-54](https://user-images.githubusercontent.com/90277741/138949443-f4c0d181-b9a4-471c-a4d0-8cab6f225ee8.png)

**Comment by elamacchia** [comment]
This is what appears when clicking on the actual image in the bucket, whereas clicking on an image uploaded with gsutil will display correctly. Additionally, downloading an errored image will just download an empty image.

**Reply by coryan** [reply]
When you upload the object with `gsutil` what is the `contentType` metadata attribute set to?  One possibility is that `gsutil` made a better guess at the actual format of the file.

The other thing is that the data is really corrupted, comparing the checksum of the uploaded object vs. the local object should answer that.

**Reply by elamacchia** [reply]
gsutil also uploads it as an image/jpeg. I did also try adding checksum verification and got the same result, where it seemingly uploaded but didn't actually get the image - gsutil is able to upload the same images, so I don't think the files are corrupted. I did try your repro script and got the same results where the image file just shows an error.

**Reply by coryan** [reply]
I think you found a bug.  The checksums and size for the file uploaded via the C++ client library do not match.

**Reply by elamacchia** [reply]
in that case, how was your script able to upload an image?

**Reply by coryan** [reply]
Most likely I just misread the results.

**Comment by elamacchia** [comment]
But you were able to view the image in the cloud storage?
________________________________

**Reply by coryan** [reply]
I saw *an* image. I picked some random jpeg file, did not care (or checked) what it looked like, just that it would not "error".  And what I got was some blank square and I just assumed that were the contents.  Why you got an error and I got some random blank square?  No idea.

**Comment by elamacchia** [comment]
I see, thank you for clarifying, just wanted to make sure we were experiencing the same results
________________________________

**Comment by coryan** [comment]
As a workaround use `gcs::WithObjectMetadata` 

```patch
diff --git a/quickstart.cc b/quickstart.cc
index 6df793a..8ab34b9 100644
--- a/quickstart.cc
+++ b/quickstart.cc
@@ -1,7 +1,7 @@
 #include <google/cloud/storage/client.h>
 #include <iostream>
 
-int main(int argc, char* argv[]) try {
+int main(int argc, char *argv[]) try {
   if (argc != 3) {
     std::cerr << "Missing bucket name or image filename.\n";
     std::cerr << "Usage: quickstart <image-file> <bucket-name>\n";
@@ -16,12 +16,17 @@ int main(int argc, char* argv[]) try {
   // Create a client to communicate with Google Cloud Storage. This client
   // uses the default configuration for authentication and project id.
   auto client = gcs::Client();
-  auto metadata = client.UploadFile(image_filename, bucket_name, image_filename, gcs::ContentType("image/jpeg")).value();
+  auto metadata =
+      client
+          .UploadFile(image_filename, bucket_name, image_filename,
+                      gcs::WithObjectMetadata(
+                          gcs::ObjectMetadata{}.set_content_type("image/jpeg")))
+          .value();
 
   std::cout << "Image uploaded: " << metadata << "\n";
 
   return 0;
-} catch (std::exception const& ex) {
+} catch (std::exception const &ex) {
   std::cerr << "Standard C++ exception thrown: " << ex.what() << std::endl;
   return 1;
 }
```

**Reply by coryan** [reply]
I merged the fixes for this last night, there will be a new release next week if you need to wait for that.

**Comment by coryan** [comment, marked as answer]
In case anybody finds this later, this is a bug, continued discussion in #7520
